WordPress/gutenberg
# PaletteEdit: Consider digits when generating kebab-cased slug

Fixes #55587
More context here: https://github.com/WordPress/gutenberg/issues/55587#issuecomment-1835443440

## What?
This PR makes the `PaletteEdit` component take numbers into account when automatically generating the `slug` converted to kebab cases based on the palette name.

As a result, for example, when you enter the palette name `MyColor1`, the generated slug will change from `my-color1` to `my-color-1`.

## Why?

If you create a custom color, a CSS variable will be generated based on that slug. When you apply that custom color to the site, that CSS variable is referenced as the value.

```css
body {
    --wp--preset--color--custom-my-color-1: #000;
}

body {
	background-color: var(--wp--preset--color--custom-my-color1);
}
```

Notice that the CSS variables are different.

The CSS variable name is `my-color-1` because [the _wp_to_kebab_case() function is used](https://github.com/WordPress/gutenberg/blob/f987124f44e5d22d64a8e232759f5a213c329bc8/lib/class-wp-theme-json-gutenberg.php#L1815) when the CSS variable is generated on the server side, and numbers are also taken into account when converting to kebab case.

But the CSS variable applied as value is `color1`. When generating a slug from a name in the Global Style custom color palette, we use the `paramCase()` ( as `kebabCase()` ) function from the `change-case` package, but this function does not take numbers into account.

## How?

We need to unify the kebab-cased slug, with numbers taken into account.

I think this process needs to be done as early as possible. Before the global style entity is saved, i.e. when generating the slug in the palette. This is because custom colors are stored in the database as values like the following, so it is difficult to extract which string is the real slug.

```
var:preset|color|custom-my-color1
```

## Testing Instructions

- If you have customized global styles, please reset them first.
- Go to the Site Editor > Editor Canvas > Global Styles > Colors > Palette.
- Create a palette with the name `MyColor1`.
- Apply the color to the entire site or specific blocks.
- Save the Global Style.
  - trunk: The applied color should disappear.
  - This PR: Applied colors should be preserved.
- Open the front page.
  - trunk: No color should be applied.
  - This PR: The intended color should be applied

## Screenshots or screencast 


https://github.com/WordPress/gutenberg/assets/54422211/b247f946-52c0-40e3-9b7a-0ef86c132b14
